File: 07Binary Search_실전/고정점 찾기.cpp
#include <iostream>
#define SIZE 1000001
using namespace std;

int n;
int* nums;

int Binary_search(int* nums, int start, int end) {
	if (start > end) return -1;

	int mid = (start + end) / 2;

	if (nums[start] == mid)
		return mid;

	else if (nums[mid] > mid) // 오른쪽에 있으면
		return Binary_search(nums, start, mid - 1);
	else
		return Binary_search(nums, mid + 1, end);
}

int main(void) {
	cin >> n;
	nums = new int[SIZE]; //동적할당 시작
	for (int i = 0; i < n; i++) {
		cin >> nums[i];
	}
	int idx = Binary_search(nums, 0, n - 1);

	cout << idx << endl;
	return 0;
}

File: 07Binary Search_실전/고정점  찾기.py
n=int(input())
nums=list(map(int,input().split()))

#이진 탐색
def Binary_search(nums,start,end):
    #존재하지 않는 경우
    if start > end:
        return None

    mid = (start + end)//2
    if nums[start] == mid: #찾은 경우
        return mid

    elif nums[mid] > mid:
        return Binary_search(nums,start,mid-1)
    else:
        return Binary_search(nums,mid+1,end)

idx=Binary_search(nums,0,n-1)
print(idx if idx>=0 else -1)